### PR TITLE
[ALLUXIO-1756] Add unit tests for Netty data server

### PR DESCRIPTION
**Pull request by aaudiber**
https://alluxio.atlassian.net/browse/ALLUXIO-1756

This change also creates interfaces for AlluxioWorker, BlockWorker, and FileSystemWorker.

### CODE REVIEW

**Comment by aaudiber** [comment]
Jenkins, test this please

File: core/server/src/main/java/alluxio/worker/AlluxioWorkerService.java [hunk 1]
```diff
@@ -22,20 +22,18 @@
...
+   * Factory for acquiring an AlluxioWorker.
    */
-  public class Factory {
-    private static AlluxioWorkerService sAlluxioWorker = null;
+  public static class Factory {
+    private static AlluxioWorkerService sAlluxioWorker;
 
     /**
-     * @return the default Alluxio worker instance
+     * Returns a the default Alluxio worker.
```
Comment (gpang) [inline core/server/src/main/java/alluxio/worker/AlluxioWorkerService.java]: "a the" -> "the"

File: core/server/src/main/java/alluxio/worker/block/BlockWorker.java [hunk 2]
```diff
@@ -11,215 +11,33 @@
...
-  /** Block store delta reporter for master heartbeat. */
-  private BlockHeartbeatReporter mHeartbeatReporter;
-  /** Metrics reporter that listens on block events and increases metrics counters. */
-  private BlockMetricsReporter mMetricsReporter;
-  /** Session metadata, used to keep track of session heartbeats. */
-  private Sessions mSessions;
-  /** Block Store manager. */
-  private BlockStore mBlockStore;
-
+public interface BlockWorker extends Worker {
```
Comment (gpang) [inline core/server/src/main/java/alluxio/worker/block/BlockWorker.java]: Is this basically the `interface` for `DefaultBlockWorker`?

File: core/server/src/main/java/alluxio/worker/block/DefaultBlockWorker.java [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+ * This includes:
+ *
+ * Servers: {@link BlockWorkerClientServiceHandler} (RPC Server)
+ *
+ * Periodic Threads: {@link BlockMasterSync} (Worker to Master continuous communication)
+ *
+ * Logic: {@link DefaultBlockWorker} (Logic for all block related storage operations)
+ */
+@NotThreadSafe // TODO(jiri): make thread-safe (c.f. ALLUXIO-1624)
+public final class DefaultBlockWorker extends AbstractWorker implements BlockWorker {
```
Comment (gpang) [inline core/server/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: Is this basically copied from the old `BlockWorker`?

File: core/server/src/main/java/alluxio/worker/netty/UnderFileSystemDataServerHandler.java [hunk 4]
```diff
@@ -40,7 +40,7 @@
  * This class handles filesystem data server requests.
  */
 @NotThreadSafe
-public class UnderFileSystemDataServerHandler {
+class UnderFileSystemDataServerHandler {
```
Comment (gpang) [inline core/server/src/main/java/alluxio/worker/netty/UnderFileSystemDataServerHandler.java]: The other data server handlers are `final`. Is it possible for this to be final?

File: core/server/src/test/java/alluxio/worker/netty/NettyDataServerTest.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.net.InetSocketAddress;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Unit tests for {@link NettyDataServer}. The entire alluxio.worker.netty package is treated as one
+ * unit for the purposes of these tests; all classes except for NettyDataServer are package-private.
+ */
+public final class NettyDataServerTest {
```
Comment (gpang) [inline core/server/src/test/java/alluxio/worker/netty/NettyDataServerTest.java]: are there any non-success test cases?

**Comment by gpang** [comment]
@aaudiber Thanks for this improvement! I left a few comments.

File: core/server/src/main/java/alluxio/worker/block/BlockWorker.java (hunk 2 above)
Comment (aaudiber) [inline core/server/src/main/java/alluxio/worker/block/BlockWorker.java]: Yes, all I did was move the signature and javadoc to the interface

File: core/server/src/test/java/alluxio/worker/netty/NettyDataServerTest.java (hunk 5 above)
Comment (aaudiber) [inline core/server/src/test/java/alluxio/worker/netty/NettyDataServerTest.java]: added

File: core/server/src/main/java/alluxio/worker/netty/UnderFileSystemDataServerHandler.java (hunk 4 above)
Comment (aaudiber) [inline core/server/src/main/java/alluxio/worker/netty/UnderFileSystemDataServerHandler.java]: done, thanks

File: core/server/src/main/java/alluxio/worker/block/DefaultBlockWorker.java (hunk 3 above)
Comment (aaudiber) [inline core/server/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: Yes, same for `FileSystemWorker` and `DefaultFileSystemWorker`

File: core/server/src/main/java/alluxio/worker/AlluxioWorkerService.java (hunk 1 above)
Comment (aaudiber) [inline core/server/src/main/java/alluxio/worker/AlluxioWorkerService.java]: Fixed, thanks

File: core/server/src/main/java/alluxio/worker/netty/BlockDataServerHandler.java [hunk 6]
```diff
@@ -80,14 +79,10 @@ void handleBlockReadRequest(final ChannelHandlerContext ctx, final RPCBlockReadR
     final long lockId = req.getLockId();
     final long sessionId = req.getSessionId();
 
-    BlockReader reader;
+    BlockReader reader = null;
     DataBuffer buffer;
     try {
       reader = mWorker.readBlockRemote(sessionId, blockId, lockId);
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/worker/netty/BlockDataServerHandler.java]: @calvinjia / @gpang can you take a look at my change to exception handling here? Previously we would swallow any exception thrown by `readBlockRemote`. This doesn't seem right so I changed it, but want to make sure it wasn't intentional

**Comment by aaudiber** [comment]
jenkins, test this please

File: core/server/src/main/java/alluxio/worker/netty/BlockDataServerHandler.java (hunk 6 above)
Comment (gpang) [inline core/server/src/main/java/alluxio/worker/netty/BlockDataServerHandler.java]: It was introduced in this commit: https://github.com/Alluxio/alluxio/commit/0544462b09b77138a48f4545e7702daeec6967a8

but I don't remember why. Maybe @uronce-cc might know?

Comment (aaudiber) [inline core/server/src/main/java/alluxio/worker/netty/BlockDataServerHandler.java]: @calvinjia ?

**Comment by gpang** [comment]
LGTM

File: core/server/src/main/java/alluxio/worker/netty/BlockDataServerHandler.java (hunk 6 above)
Comment (calvinjia) [inline core/server/src/main/java/alluxio/worker/netty/BlockDataServerHandler.java]: I think the new logic is correct. We probably also want to do `req.validate` before anything else.

Comment (aaudiber) [inline core/server/src/main/java/alluxio/worker/netty/BlockDataServerHandler.java]: I've moved `req.validate` to the start

**Comment by aaudiber** [comment]
jenkins, test this please

**Pull request merged**